File: rust/src/problems/find_and_replace_pattern.rs
use std::collections::{HashMap, HashSet};
use std::iter::FromIterator;

pub fn find_and_replace_pattern(words: Vec<String>, pattern: String) -> Vec<String> {
    return words
        .into_iter()
        .filter(|w| is_same_pattern(w.parse().unwrap(), pattern.clone()))
        .collect();
}

fn is_same_pattern(word: String, pattern: String) -> bool {
    let mut hash: HashMap<char, char> = HashMap::new();
    for (p, w) in word.chars().into_iter().zip(pattern.chars().into_iter()) {
        if hash.insert(p, w).unwrap_or(w) != w {
            return false;
        }
    }
    return HashSet::<&char>::from_iter(hash.values()).len() == hash.values().len();
}

File: go/problems/find_and_replace_pattern.go
package problems

//https://leetcode.com/problems/find-and-replace-pattern/
func FindAndReplacePattern(words []string, pattern string) []string {
	var results []string
	for _, word := range words {
		if isSamePattern(word, pattern) {
			results = append(results, word)
		}
	}
	return results
}

func isSamePattern(word string, pattern string) bool {
	hash := make(map[rune]rune)

	if len(word) != len(pattern) {
		return false
	}
	isValid := true
	for index, char := range pattern {
		if _, ok := hash[char]; !ok {
			if includes(hash, rune(word[index])) {
				isValid = false
				break
			}
			hash[char] = rune(word[index])
		} else if rune(word[index]) != hash[char] {
			isValid = false
			break
		}
	}
	return isValid
}

func includes(hash map[rune]rune, char rune) bool {
	for _, v := range hash {
		if char == v {
			return true
		}
	}
	return false
}
